### PR TITLE
Type hint on input x gradient

### PR DESCRIPTION
**Pull request by luciawlli**
$ ./scripts/run_mypy.sh
360 passed, 50 skipped, 138 warnings in 106.13 seconds

$ black .
All done! ✨ 🍰 ✨
107 files left unchanged.

$ flake8 .
<no errors>

$ ./scripts/run_mypy.sh
Success: no issues found in 44 source files
Success: no issues found in 48 source files

### CODE REVIEW

File: tests/attr/test_input_x_gradient.py [hunk 1]
```diff
@@ -96,14 +101,16 @@ def _input_x_gradient_classification_assert(self, nt_type="vanilla"):
             output = model(input)[:, target]
             output.backward()
             expercted = input.grad * input
-            self.assertEqual(
-                expercted.detach().numpy().tolist(),
-                attributions.detach().numpy().tolist(),
-            )
+            if isinstance(attributions, Tensor):
```
Comment (luciawlli) [inline tests/attr/test_input_x_gradient.py]: added if check here and below, otherwise ./scripts/run_mypy.sh gives errors on 
```

tests/attr/test_input_x_gradient.py:87: error: Item "Tuple[Tensor, ...]" of "Union[Tensor, Tuple[Tensor, ...]]" has no attribute "shape"
tests/attr/test_input_x_gradient.py:107: error: Item "Tuple[Tensor, ...]" of "Union[Tensor, Tuple[Tensor, ...]]" has no attribute "detach"
```

due to `inputs: Union[Tensor, Tuple[Tensor, ...]]`

Comment (vivekmig) [reply tests/attr/test_input_x_gradient.py]: The change above to TensorOrTupleOfTensors should avoid this issue, and you can remove this and the extra case below.

File: captum/attr/_core/input_x_gradient.py [hunk 2]
```diff
@@ -14,7 +17,14 @@ def __init__(self, forward_func):
         """
         GradientAttribution.__init__(self, forward_func)
 
-    def attribute(self, inputs, target=None, additional_forward_args=None):
+    def attribute(
+        self,
+        inputs: Union[Tensor, Tuple[Tensor, ...]],
```
[the remark this answers is not in the document]
Comment (luciawlli) [reply captum/attr/_core/input_x_gradient.py]: Thanks for the review.

After this change, I also made the change in `test_input_x_gradient.py` line 52 to make `input TensorOrTupleOfTensors`, however, the`if isinstance ...` is still needed, else would run into the error:

`tests/attr/test_input_x_gradient.py:88: error: "Tuple[Tensor, ...]" has no attribute "shape"`

File: tests/attr/test_input_x_gradient.py [hunk 3]
```diff
@@ -75,15 +78,18 @@ def _input_x_gradient_base_assert(
                     )
                 self.assertEqual(input.shape, attribution.shape)
         else:
```
Comment (vivekmig) [inline tests/attr/test_input_x_gradient.py]: Can you change this else to `elif isinstance(attributions, Tensor) `?
This captures the only other possibility for attributions and would not need any isinstance checks below.

File: tests/attr/test_input_x_gradient.py [hunk 4]
```diff
@@ -106,4 +112,4 @@ def _input_x_gradient_classification_assert(self, nt_type="vanilla"):
                 input, nt_type=nt_type, n_samples=10, stdevs=1.0, target=target
             )
 
-        self.assertAlmostEqual(attributions.shape, input.shape)
+        assertArraysAlmostEqual(attributions.shape, input.shape)
```
Comment (vivekmig) [inline tests/attr/test_input_x_gradient.py]: Can you change this to just self.assertEqual? Thanks!

File: tests/attr/test_input_x_gradient.py [hunk 5]
```diff
@@ -1,55 +1,58 @@
...
         self,
-        model,
-        inputs,
-        expected_grads,
-        additional_forward_args=None,
-        nt_type="vanilla",
-    ):
+        model: Module,
+        inputs: TensorOrTupleOfTensors,
+        expected_grads: Tensor,
```
Comment (vivekmig) [inline tests/attr/test_input_x_gradient.py]: This should also be TensorOrTupleOfTensors

Comment (luciawlli) [reply tests/attr/test_input_x_gradient.py]: Changing `expected_grads` from `Tensor` to `TensorOrTupleOfTensors` results in the following error in 3 places:
`tests/attr/test_input_x_gradient.py:35: error: Value of type variable "TensorOrTupleOfTensors" of "_input_x_gradient_base_assert" of "Test" cannot be "object"`

**Comment by luciawlli** [comment]
> Looks great, thanks again! Just a few more small changes, and it should be good.

Thanks for reviewing! Addressed all the comments :)

File: tests/attr/test_input_x_gradient.py [hunk 6]
```diff
@@ -74,16 +77,18 @@ def _input_x_gradient_base_assert(
                         attribution.reshape(-1), (expected_grad * input).reshape(-1)
                     )
                 self.assertEqual(input.shape, attribution.shape)
-        else:
-            if nt_type == "vanilla":
+        elif isinstance(attributions, Tensor):
+            if nt_type == "vanilla" and isinstance(attributions, Tensor):
```
Comment (vivekmig) [inline tests/attr/test_input_x_gradient.py]: I think this isinstance can also be removed now?

**Comment by luciawlli** [comment]
> Thanks for updating! Just one more minor fix.

Yes, it should be cleaned up. Sorry missed this early.
